wpsc-includes/variations.class.php

```diff
@@ -296,3 +296,117 @@ function wpsc_get_child_object_in_terms_var( $parent_id, $terms, $taxonomies, $a
 		return false;
 	}
 }
+
+/**
+ * Given array of variation selections this works through the terms and returns the product_id for the matching variation
+ *
+ * For the `$variations` parameter we expect to get two 'term_ids' which correspond to the selections in the variation.
+ * So if we have a blue large shirt and the blue term_id is 8 and the large term_id is 12 we would get an array that looks like.
+ *
+ * array(
+ *      '2' => '8',     // this is the blue term_id
+ *      '9' => '12',    // this is the large term_id
+ * );
+ *
+ * Here the keys are captured when someone clicks the 'add to cart' button and correspond with ... whatever. They don't really matter for our function.
+ * Really ou could pass an array of 2 term_ids here and a $product_id that has variations to match those terms and you'd get back the expected array.
+ *
+ * @since 4.0
+ *
+ * @param array     $variations     required            The array of variation selections
+ * @param int       $product_id     required            The default product_id
+ * @uses wpsc_get_child_objects_in_terms()              Given $product_id and product params this returns the variation product id
+ * @return array    $args {
+ *      @param  int     product_id                      The variation product_id
+ *      @param  array   variation_values                The array of variation_values that wpsc_add_to_cart needs to populate $provided_parameters['variation_values']
+ * }
+ */
+function wpsc_get_product_data_from_variations( $variations, $product_id ) {
+
+	$variation_values = array();
+
+	foreach ( (array) $variations as $key => $variation ) {
+		$variation_values[ (int) $key ] = (int) $variation;
+	}
+
+	if ( count( $variation_values ) > 0 ) {
+		$variation_product_id = wpsc_get_child_object_in_terms( $product_id, $variation_values, 'wpsc-variation' );
+		if ( $variation_product_id > 0 ) {
+			$product_id = $variation_product_id;
+		}
+	}
+
+	/**
+	 * Allows users to filter the product_id based on the variation selections
+	 *
+	 * @since 4.0
+	 *
+	 * @param int   $product_id             The default passed product_id
+	 * @param array $variations             The variation selections passed to the core function
+	 */
+	$product_id = apply_filters( 'wpsc_variation_product_id', absint( $product_id ), $variations );
```

Comment:
I like that the product ID is filtered the same way in both functions, but rather than duplicating the filter, let's do it once, and reference it in the other, like [WordPress does](https://github.com/WordPress/WordPress/blob/master/wp-includes/option.php#L99-L100).

Also, let's make sure that we're able to pass the original `$product_id` passed into the function, as well as the newly retrieved `$variation_product_id`.